Frictionless and frictional contact with new mortar

lindsayad opened the pull request:
I'm getting this up so I can reference it for the [June newsletter](https://github.com/idaholab/moose/pull/13506/).

This PR demonstrates use of the new mortar system along with nonlinear complimentarity functions for solving frictionless and frictional mechanical contact. A summary of performance for a bouncing block problem is given in the tables below:

### Frictionless contact

| Lagrange multiplier | Displacement | NCP function | Time (arbitrary units) | Time steps | Nonlinear iterations |
| ------------------- | ------------ | ------------ | ---------------------- | ---------- | -------------------- |
| Nodal | Mortar | Min | 14.401 | 40 | 104 |
| Nodal | Mortar | FB | 17.752 | 40 | 135 |
| Nodal | Nodal | Min | 5.438 | 41 | 104 |
| Nodal | Nodal | FB | 6.770 | 41 | 149 |
| Mortar | Mortar | Min | 14.454 | 40 | 106 |
| Mortar | Mortar | FB | 19.027 | 40 | 136 |

Notes:

- Clearly having mortar mesh generation slows the simulation down, which is not surprising
- The min NCP function is undeniably better here in the frictionless case
- For the pure nodal algorithms, the time step that did not converge featured classic ping-ponging behavior, which can be extremely problematic for larger scale problems:

<pre>
 5 Nonlinear |R| = <b><i>4.007951e-04</b></i>
    |residual|_2 of individual variables:
                  disp_x:    0.000399808
                  disp_y:    2.75599e-05
                  normal_lm: 5.52166e-06


The number of nodes in contact is 11

      0 Linear |R| = 4.007951e-04
      1 Linear |R| = 1.287307e-04
      2 Linear |R| = 8.423398e-06
      3 Linear |R| = 1.046825e-07
      4 Linear |R| = 8.017310e-09
      5 Linear |R| = 3.053040e-10
  Linear solve converged due to CONVERGED_RTOL iterations 5
 6 Nonlinear |R| = <b><i>4.432193e-04</b></i>
    |residual|_2 of individual variables:
                  disp_x:    0.000396694
                  disp_y:    0.00019545
                  normal_lm: 2.96013e-05


The number of nodes in contact is 11

      0 Linear |R| = 4.432193e-04
      1 Linear |R| = 1.355935e-04
      2 Linear |R| = 1.216010e-05
      3 Linear |R| = 6.386952e-07
      4 Linear |R| = 2.235594e-08
      5 Linear |R| = 2.884193e-10
  Linear solve converged due to CONVERGED_RTOL iterations 5
 7 Nonlinear |R| = <b><i>4.008045e-04</b></i>
    |residual|_2 of individual variables:
                  disp_x:    0.000399816
                  disp_y:    2.76329e-05
                  normal_lm: 5.29313e-06


The number of nodes in contact is 11

      0 Linear |R| = 4.008045e-04
      1 Linear |R| = 1.287272e-04
      2 Linear |R| = 8.423081e-06
      3 Linear |R| = 1.047782e-07
      4 Linear |R| = 8.054781e-09
      5 Linear |R| = 3.046073e-10
  Linear solve converged due to CONVERGED_RTOL iterations 5
 8 Nonlinear |R| = <b><i>4.432194e-04</b></i>
</pre>


### Frictional contact

| LM normal | LM tangential | Displacement | NCP function normal | NCP function tangential | Time (arbitrary units) | Time steps | Nonlinear iterations |
| --------- | ------------  | ------------ | ------------------- | ----------------------- | ---------------------- | ---------- | -------------------- |
| Mortar | Mortar | Mortar | FB | FB | 18.771 | 43 | 190 |
| Mortar | Mortar | Mortar | min | FB | 18.281 | 41 | 194 |
| Nodal | Mortar | Mortar | min | FB | 15.544 | 41 | 165 |
| Mortar | Mortar | Mortar | Min | Min | 74.801 | 58 | 320 |
| Nodal | Nodal | Mortar | Min | Min | 36.923 | 52 | 230 |
| Nodal | Nodal | Mortar | FB | FB | NA | NA | NA |

lindsayad committed:
Run mortar-mech test with PJFNK and not FD

Give a heuristic for approximation the dependence of the gap on the
displacement variables, which is the most important missing piece of the
Jacobian. Other things like the dependence of tests, normals, Jxw, phis,
etc. on the displacements are not included, so we still need to use PJFNK.
The Jacobian norm ratio is approximately 1e-2 to 1e-3, so I think that's
pretty good. With lu as the preconditioner, the linear iteration count never
exceeded 1-2.

lindsayad committed:
Make sure to reinit ghosting if we have displaced mesh objects

lindsayad committed:
Add tangential contact

lindsayad committed:
Add tangential contact test

lindsayad committed:
Add production test of mortar LM alongside penalty frictional sliding block test

lindsayad committed:
Add nodal NCP version

lindsayad committed:
Create ContactDofSet postprocessor

lindsayad committed:
In modules, add NodalNormalContactLM and reorganize other lm-based constraints

lindsayad committed:
Add more mortar tests in moose test

Includes:
- Nodal enforcement of contact constraints
    - Nodal enforcement is more accurate (exact constraint enforcement at the nodes)
      and more performant in terms of non-linear iterations and the unix time utility
      than mortar enforcement of the constraints
- Different NCP functions for mortar enforcement of contact constraints
    - The std::min NCP function is more performant in terms of non-linear iterations
      than the FB function

lindsayad committed:
Begin creating nodal tangential lm constraint

lindsayad committed:
Organizing and documentation different algorithms performance

lindsayad committed:
WIP on modules test input files

lindsayad committed:
Finished logging frictionless algorithm results

Now also working on improving frictional contact by smoothing some of
the discontinuous functions in the NodalTangentialContactLM class. One
consequence of smoothing is introduction of non-linearity in regions away
from the discontinuity, so we'll see whether we make the problem worse instead of
better

lindsayad committed:
Finish documenting frictionless/frictional cases and generate gold files

lindsayad committed:
Add most promising frictional algorithm yet

Refs #13080

lindsayad committed:
pragma once

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/346566/) on 909076c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13508/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13508/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c3c0499a28863c022c28b01f8904287617878b42`

lindsayad committed:
Delete lm_constraints directories

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/346592/) on d90ee76 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13508/site)

This comment will be updated on new commits.

lindsayad committed:
Consistently rename all contact objects

lindsayad committed:
Add frictionless lm sliding block test in combined

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/347001/) on 42c1203 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13508/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13508/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5d1ad78212fb38b837fd75f7972a948bcb1e8fe1`

lindsayad committed:
Ensure that the tangential lagrange multiplier acts in the correct direction

lindsayad committed:
Consistently use the 'c' parameter

lindsayad committed:
Apparently removing the std::abs in application of the tangential forces changes the results...

lindsayad committed:
Complete sliding_block frictional test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/347265/) on ed897e3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13508/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13508/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5d1ad78212fb38b837fd75f7972a948bcb1e8fe1`

lindsayad committed:
Add predictor to frictional lm test to make it fly even faster

lindsayad committed:
Throw exceptions instead of erroring out

When we get a bad search in NearestNodeThread we should throw an exception
and decrease the time step instead of erroring out

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/347282/) on e2bafba wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13508/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13508/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5d1ad78212fb38b837fd75f7972a948bcb1e8fe1`

lindsayad commented:
@novasr Here's a comparison of the mortar frictional input file I've added to the sliding block test directory vs the corresponding penalty input file:

| Algorithm | Time (arb. units) | Time steps to end time | Cumulative non-linear iterations |
| --- | --- | --- | --- |
| Mortar LM | 13.901 | 151 | 476 |
| Penalty | 20.711 | 151 | 938 |

There's a cost associated with generation of the mortar segment mesh so that partially offsets the fact that the mortar case takes nearly half the non-linear iterations of the penalty case. However, my hope is that where mortar can really shine is in some of the really difficult frictional contact assessment cases.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/347601/) on c7a0245 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13508/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13508/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5d1ad78212fb38b837fd75f7972a948bcb1e8fe1`

lindsayad committed:
Explicitly qualify with this

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/347652/) on 684a97f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13508/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13508/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5d1ad78212fb38b837fd75f7972a948bcb1e8fe1`

lindsayad committed:
Documentation

lindsayad committed:
Use snes_ksp_ew for troublesome solves

lindsayad committed:
Use snes_ksp_ew for all frictional cases

lindsayad committed:
Transfer results into moose docs

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/348558/) on cdc65f9 : invalidated by @lindsayad

tophmatthews commented:
This is the last one for the trifecta !!! (AD, scaling, mortar method)

lindsayad commented:
Yea, not going to let @bwspenc give this one the kiss of death with his rapid PR reviewing skills 😄 If I don't hear anything soon, I'll have someone else look at it.

tophmatthews commented:
(ref #11562 #12700) :)

Nah, actually I have faith @bwspenc will prove us wrong!

bwspenc commented:
I will prove you wrong.  I'm going to have to push this off to next week though, because I'm busily working on the other massive PR that I've been reviewing way too long...

lindsayad commented:
T minus two days :smile:

dschwen commented on framework/doc/content/source/postprocessors/ContactDofSet.md:
```diff
@@ -0,0 +1,22 @@
+# ContactDofSet
```
This should probably be named `ContactDOFSetSize` or  `ContactDOFNumber` to indicate that it outputs a number rather than a set.

dschwen commented on framework/doc/content/syntax/Constraints/index.md:
```diff
@@ -2,22 +2,36 @@
 An excellent overview of the conservative mortar constraint implementation in
 MOOSE is given in
-[this technical report](https://www.osti.gov/biblio/1468630/). There are five
+[this technical report](https://www.osti.gov/biblio/1468630/). We have verified
```
use `[cite!...]` maybe?

dschwen commented on framework/include/base/Assembly.h:
```diff
@@ -258,6 +258,12 @@ class Assembly
+   * Returns the array of tangents for quadrature points on a current side
+   * @return A _reference_.  Make sure to store this as a reference!
+   */
+  const MooseArray<std::vector<Point>> & tangents() const { return _current_tangents; }
```
Are these guaranteed to be an orthogonal basis of the tangent space?

lindsayad replied:
Yes. These actually come out of libMesh

dschwen reviewed (commented):
The bits that I understand look good to me :-D

lindsayad committed:
ContactDofSet -> ContactDOFSetSize

lindsayad commented:
@idaholab/moose-team Could I get a reviewer on this? One issue that I didn't know how to tackle is this:

I really like the Contact mortar constraints for testing the mortar system, which is a framework capability. But the contact constraints are very physicsy, and we have a Contact module 😆 I guess maybe my existing mortar constraint tests provide enough coverage and I should just move those constraint objects and this round of tests into the Contact module...

lindsayad committed:
Move from framework into contact module

lindsayad commented:
Ok, I went ahead and moved the mechanical contact code into the `Contact` module

permcody commented:
@lindsayad - I'll give this a review tomorrow

fdkong commented on modules/contact/src/constraints/NormalNodalLMMechanicalContact.C:
```diff
@@ -0,0 +1,204 @@
+  _qp = 0;
+
+  re(0) = computeQpResidual(Moose::Slave);
+}
```
It looks like you are not supporting the tagging capability. It would be great, if we could add that.

lindsayad replied:
Slack pack not doing that :smile: Someday

fdkong commented on modules/contact/src/constraints/NormalNodalLMMechanicalContact.C:
```diff
@@ -0,0 +1,204 @@
+
+  for (_j = 0; _j < _phi_master.size(); ++_j)
+    Ken(0, _j) += computeQpOffDiagJacobian(Moose::SlaveMaster, jvar);
+}
```
Similarly for all these routines.

fdkong commented on modules/contact/src/constraints/NormalNodalLMMechanicalContact.C:
```diff
@@ -0,0 +1,204 @@
+      Real b = _u_slave[_qp];
+
+      if (_ncp_type == "fb")
+        return a + b - std::sqrt(a * a + b * b + _epsilon);
```
Nice

fdkong commented on modules/contact/src/constraints/NormalNodalLMMechanicalContact.C:
```diff
@@ -0,0 +1,204 @@
+        return (a + b - std::sqrt(a * a + b * b + _epsilon)).derivatives();
+      else
+        return std::min(a, b).derivatives();
+    }
```
`a` and `b` are always positive?

lindsayad replied:
The std::min function will ensure that either a or b will be zero, and consequently that the other will be positive

fdkong commented on modules/contact/test/src/constraints/NormalMortarLMMechanicalContact.C:
```diff
@@ -0,0 +1,131 @@
+              _lambda[_qp] + gap - std::sqrt(_lambda[_qp] * _lambda[_qp] + gap * gap + _epsilon);
+        else
+          fb_function = std::min(_lambda[_qp], gap);
+
```
Minor thing. If we still have `a = _lambda[_qp], and b = gap`,  it will be consistent with other part of the code. It is easier to understand for me.

lindsayad replied:
Done

fdkong commented on modules/contact/test/src/constraints/NormalMortarLMMechanicalContact.C:
```diff
@@ -0,0 +1,131 @@
+          fb_function =
+              _lambda[_qp] + gap - std::sqrt(_lambda[_qp] * _lambda[_qp] + gap * gap + _epsilon);
+        else
+          fb_function = std::min(_lambda[_qp], gap);
```
Simliar.

lindsayad replied:
Done

fdkong commented on modules/contact/test/tests/bouncing-block-contact/frictional-mortar-fb-lm-mortar-disp.i:
```diff
@@ -0,0 +1,219 @@
+  solve_type = 'PJFNK'
+  petsc_options = '-snes_converged_reason -ksp_converged_reason -pc_svd_monitor -snes_linesearch_monitor -snes_ksp_ew'
+  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount -mat_mffd_err'
+  petsc_options_value = 'lu       NONZERO               1e-15                   1e-5'
```
Anything rather than LU will work as well?

lindsayad replied:
Yes, but the performance is quite a bit worse, perhaps because of the lagrange multiplier which introduces a saddle point. For another input file in this directory, I get 152 non-linear iterations with lu in serial, 416 non-linear iterations with ilu in serial, and 391 non-linear iterations with asm and ilu for the sub pc with two processes.

I would like to try out automatic scaling on these problems once I've introduced another option other than diagonal scaling.

lindsayad committed:
NCP a,b function consistency

permcody commented:
@lindsayad - needs a rebase

lindsayad committed:
Update to new getParam API

lindsayad commented:
Alright, rebase done

lindsayad commented:
If this gets merged, I will actually start writing a paper on it...

fdkong commented:
> If this gets merged, I will actually start writing a paper on it...

Done

lindsayad commented:
Sweet!

On Mon, Jul 1, 2019 at 11:17 AM Fande Kong <notifications@github.com> wrote:

> If this gets merged, I will actually start writing a paper on it...
>
> Done
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13508?email_source=notifications&email_token=ACOGA4ACSK3MOP5PL7NTBKLP5JC33A5CNFSM4HSI5AS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY6522Q#issuecomment-507370858>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4BLYTPKISB6SLW23F3P5JC33ANCNFSM4HSI5ASQ>
> .
>

fdkong commented:
Something failings at next https://civet.inl.gov/job/360929/

lindsayad commented:
K, I'll figure out how to address it.

On Mon, Jul 1, 2019 at 11:56 AM Fande Kong <notifications@github.com> wrote:

> Something failings at next https://civet.inl.gov/job/360929/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13508?email_source=notifications&email_token=ACOGA4HN7OVPPW2QOWNUNP3P5JHMRA5CNFSM4HSI5AS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY7A7DA#issuecomment-507383692>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4FM7QVZCFW54CCJ3YDP5JHMRANCNFSM4HSI5ASQ>
> .
>

lindsayad commented:
@novasr Now that this is in, let me know if you want me to run any BISON cases.

tophmatthews commented:
It may be good to wrap this up in an action too?

lindsayad commented:
Seems like that might be good for someone else to do to get some knowledge transfer

tophmatthews commented:
Yeah, that's fair. I can do it, but not for another month

novasr commented:
Great idea for someone to write the action for knowledge transfer @lindsayad.  

Try running this case with friction:
bison/assessment/metallic_fuel/EBRII/X441/analysis/group_F/x441_base.i

I believe this is the case with 85% smear density, and we've had trouble running it with friction. @acasagran confirm.

dschwen commented:
This is follow on work for our Bison robustness and performance effort. I will work on it.

lindsayad commented:
Work on what @dschwen? Writing the action or running the assessment case?
You can do both, but I'm also going to run the assessment case because I
want to see what's going to happen! :-)

On Tue, Jul 2, 2019 at 6:03 AM Daniel Schwen <notifications@github.com>
wrote:

> This is follow on work for our Bison robustness and performance effort. I
> will work on it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13508?email_source=notifications&email_token=ACOGA4C5MQOLQ5X26CPEQBLP5NG2XA5CNFSM4HSI5AS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBGGZQ#issuecomment-507667302>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4FWMIB7ZFOXDI526FLP5NG2XANCNFSM4HSI5ASQ>
> .
>

dschwen commented:
Action. So, should we maybe add this as an option to the old action?

lindsayad commented:
Probably if we want to make user adoption as simple as possible

On Tue, Jul 2, 2019 at 7:36 AM Daniel Schwen <notifications@github.com>
wrote:

> Action. So, should we maybe add this as an option to the old action?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13508?email_source=notifications&email_token=ACOGA4FKXB3AZZVC55MKPQLP5NRW5A5CNFSM4HSI5AS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBPUIY#issuecomment-507705891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4AGIANXYSNMH3KQM4LP5NRW5ANCNFSM4HSI5ASQ>
> .
>

acasagran commented:
Yes. Group F is the case from the X441 models that has the 85% smear density and will get into contact fairly soon during the analysis. I've only gotten frictionless contact to work. Glued and frictional contact fail pretty quickly after contact is established I think due to problems with the fuel creep material model.

tophmatthews commented:
This needs to be worked into thermal contact as well. I cleaned it up what is currently in there as best I could earlier this year, but it'll probably take some work.

tophmatthews commented:
> If this gets merged, I will actually start writing a paper on it...

@lindsayad?

lindsayad commented:
Haha no I haven't started it yet... I'll move it forward on the agenda!

tophmatthews commented:
Maybe I can get you some badass simulations of the application if you need them

tophmatthews commented:
Really, I just wanted to know where to give you credit on getting and other AD/AS things in.

lindsayad commented:
Well that would actually be really beneficial. So far I have toy problem results in the test suites that show this thing can work better than our preexisting models, but I haven't had time to do any real physics cases yet.

lindsayad commented:
Sadly there are no publications yet that include AD/AS/mortar. We just submitted a manuscript to SoftwareX though

tophmatthews commented:
With the FY finally out of the way, I'm going full steam ahead on this, but I obviously [need help](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/moose-users/JA1yt77MEV0) along the way.
